e2e/rapid_test.go

```diff
@@ -0,0 +1,573 @@
+	if bf.IsStuckMock != nil {
+		return bf.IsStuckMock(num)
+	}
+	panic("IsStuck " + strconv.Itoa(int(num)))
+}
+
+func (bf *BackendFake) ValidateCommit(from pbft.NodeID, seal []byte) error {
+	return nil
+}
+
+func TestPropertySeveralHonestNodesCanAchiveAgreement(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		numOfNodes := rapid.IntRange(4, 30).Draw(t, "num of nodes").(int)
+		ft := &fakeTransport{}
+		cluster, timeoutsChan := generateCluster(numOfNodes, ft)
+		for i := range cluster {
+			cluster[i].SetInitialState(context.Background())
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+
+		err := runCluster(ctx,
+			cluster,
+			sendTimeoutIfNNodesStucked(t, timeoutsChan, numOfNodes),
+			func(doneList *BoolSlice) bool {
+				//everything done. All nodes in done state
+				if doneList.CalculateNum(true) == numOfNodes {
+					return true
+				}
+				return false
+			}, func(maxRound uint64) bool {
+				//something went wrong.
+				if maxRound > 3 {
+					t.Error("Infinite rounds")
+					return true
+				}
+				return false
+			}, 100)
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+}
+
+func TestPropertySeveralNodesCanAchiveAgreementWithFailureNodes(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		numOfNodes := rapid.IntRange(4, 30).Draw(t, "num of nodes").(int)
+		routingMapGenerator := rapid.MapOfN(
+			rapid.Uint64Range(0, uint64(numOfNodes)-1),
+			//not used
+			rapid.Bool(),
+			2*numOfNodes/3+1,
+			numOfNodes-1,
+		).Filter(func(m map[uint64]bool) bool {
+			_, ok := m[0]
+			return ok
+		})
+
+		routingMap := routingMapGenerator.Draw(t, "generate routing").(map[uint64]bool)
+		ft := &fakeTransport{
+			GossipFunc: func(ft *fakeTransport, msg *pbft.MessageReq) error {
+				from, err := strconv.ParseUint(string(msg.From), 10, 64)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if _, ok := routingMap[from]; ok {
+					for i := range routingMap {
+						if i == from {
+							continue
+						}
+						ft.nodes[i].PushMessage(msg)
+					}
+				}
+
+				return nil
+			},
+		}
+		cluster, timeoutsChan := generateCluster(numOfNodes, ft)
+		for i := range cluster {
+			cluster[i].SetInitialState(context.Background())
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+
+		err := runCluster(ctx,
+			cluster,
+			sendTimeoutIfNNodesStucked(t, timeoutsChan, numOfNodes),
+			func(doneList *BoolSlice) bool {
+				//check that 3 node switched to done state
+				if doneList.CalculateNum(true) >= numOfNodes*2/3+1 {
+					//everything done. Success.
+					return true
+				}
+				return false
+			}, func(maxRound uint64) bool {
+				if maxRound > 10 {
+					t.Error("Infinite rounds")
+					return true
+				}
+				return false
+			}, 100)
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+}
+
+func TestProperty4NodesCanAchiveAgreementIfWeLockButNotCommitProposer_Fails(t *testing.T) {
+	t.Skip("Unskip when fix")
```

Comment:
This is skipped due to the issue on the consensus algorithm (one of the liveness issues I believe; looks like the one where there is not quorum of commit messages in order to insert a proposal)? Or the test needs to be re-worked?